### PR TITLE
Support loading RSAPSS public keys with parameters

### PR DESCRIPTION
**Pull request by gautierdelorme**
Support loading RSA PSS public keys with parameters.

### Checklist
- [x] I've run tests to see all new and existing tests pass
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](CONTRIBUTING.md)
- [x] I've updated the documentation if necessary

#### If you've made changes to `gyb` files
- [ ] I've run `.script/generate_boilerplate_files_with_gyb` and included updated generated files in a commit of this pull request

### Motivation:

Trying to load those keys currently fails.

### Modifications:

Parameters are stripped and the key is treated as a regular public key.

### Result:

`_RSA.Signing.PublicKey()` works with RSA PSS keys with parameters.

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -84,7 +84,9 @@ let package = Package(
             .library(name: "CCryptoBoringSSL", type: .static, targets: ["CCryptoBoringSSL"]),
             MANGLE_END */
     ],
-    dependencies: [],
+    dependencies: [
+        .package(url: "https://github.com/apple/swift-asn1.git", .upToNextMajor(from: "1.2.0"))
```
Comment (Lukasa) [inline Package.swift:88]: Nit:

```suggestion
        .package(url: "https://github.com/apple/swift-asn1.git", from: "1.2.0")
```

**Comment by Lukasa** [comment]
@swift-server-bot add to allowlist

File: Sources/_CryptoExtras/Util/SubjectPublicKeyInfo.swift [hunk 2]
```diff
@@ -0,0 +1,116 @@
+//
+//  SubjectPublicKeyInfo.swift
+//
+//
+//  Created by Gautier Delorme on 24/09/2024.
+//
```
Comment (Lukasa) [inline Sources/_CryptoExtras/Util/SubjectPublicKeyInfo.swift:6]: Can you add the license header that is present in other files and remove this header?

File: Sources/_CryptoExtras/RSA/RSA.swift [hunk 3]
```diff
@@ -95,9 +96,12 @@ extension _RSA.Signing {
         ///
         /// This constructor supports key sizes of 1024 bits or more. Users should validate that key sizes are appropriate
         /// for their use-case.
+        /// Parameters from RSA PSS keys will be stripped.
         /// - Warning: Key sizes less than 2048 are not recommended and should only be used for compatibility reasons.
         public init<Bytes: DataProtocol>(unsafeDERRepresentation derRepresentation: Bytes) throws {
-            self.backing = try BackingPublicKey(derRepresentation: derRepresentation)
+            let sanitizedDer = try SubjectPublicKeyInfo.stripRsaPssParameters(derEncoded: [UInt8](derRepresentation))
```
Comment (Lukasa) [inline Sources/_CryptoExtras/RSA/RSA.swift:102]: Not a request for changes in this PR, but can you file an issue that asks us to move all our RSA key parsing into Swift ASN1? This is a pragmatic change for now, but as we're _already_ parsing the key we should probably just stick with that now.

File: Sources/_CryptoExtras/Util/SubjectPublicKeyInfo.swift [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+extension SubjectPublicKeyInfo {
+    static func stripRsaPssParameters(derEncoded: [UInt8]) throws -> [UInt8] {
+        guard var spki = try? SubjectPublicKeyInfo(derEncoded: derEncoded) else {
+            // If it's not a SPKI then it can't be a PSS key so we just return it.
+            return derEncoded
+        }
+
+        if spki.algorithmIdentifier.algorithm == .AlgorithmIdentifier.rsaPSS {
```
Comment (Lukasa) [inline Sources/_CryptoExtras/Util/SubjectPublicKeyInfo.swift:107]: This doesn't look like it'll compile: I don't think that algorithm identifier is actually defined anywhere in this code.

**Comment by Lukasa** [comment]
Thanks for this PR! I've left a few notes in the diff for cleaning up.

File: Sources/_CryptoExtras/Util/SubjectPublicKeyInfo.swift (hunk 4 above)
Comment (gautierdelorme) [reply Sources/_CryptoExtras/Util/SubjectPublicKeyInfo.swift:107]: It comes from `swift-asn1`: https://github.com/apple/swift-asn1/blob/1a80ac7e4b8831264b2964758732a63265457a58/Sources/SwiftASN1/Basic%20ASN1%20Types/ObjectIdentifier.swift#L259-L263.

File: Sources/_CryptoExtras/RSA/RSA.swift (hunk 3 above)
Comment (gautierdelorme) [reply Sources/_CryptoExtras/RSA/RSA.swift:102]: https://github.com/apple/swift-crypto/issues/269

File: Sources/_CryptoExtras/Util/SubjectPublicKeyInfo.swift [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
```
Comment (Lukasa) [inline Sources/_CryptoExtras/Util/SubjectPublicKeyInfo.swift:114]: Can we turn this into a shortcut early return? If we don't have to make any modifications, we don't need to re-encode the key.

**Review by Lukasa — APPROVED** [review]
Cool, this LGTM.

**Comment by Lukasa** [comment]
Thanks very much!

**Review by Lukasa — APPROVED** [review]

**Pull request merged**